src/crypto/tls/ech.go

```diff
@@ -0,0 +1,874 @@
+					if _, ok = outer[ext]; ok {
+						b.AddUint16(ext)
+						b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+							b.AddBytes(extData)
+						})
+					}
+					return true
+				})
+				if !ok {
+					panic(cryptobyte.BuildError{Err: errReadFailure})
+				}
+			} else {
+				b.AddUint16(ext)
+				b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+					b.AddBytes(extData)
+				})
+			}
+		}
+	})
+
+	decompressedData, err := b.Bytes()
+	if err == errReadFailure {
+		return nil, nil, false // Reading failed
+	} else if err != nil {
+		panic(err) // Writing failed
+	}
+
+	return decompressedData, innerDigest, true
+}
+
+// processClientHelloExtensions interprets data as a ClientHello and applies a
+// function proc to each extension. Returns a bit indicating whether parsing
+// succeeded.
+func processClientHelloExtensions(data []byte, proc func(ext uint16, extData cryptobyte.String) bool) bool {
+	_, extensionsData, ok := splitClientHelloExtensions(data)
+	if !ok {
+		return false
+	}
+
+	s := cryptobyte.String(extensionsData)
+	if s.Empty() {
+		return true
+	}
+
+	var extensions cryptobyte.String
+	if !s.ReadUint16LengthPrefixed(&extensions) || !s.Empty() {
+		return false
+	}
+
+	for !extensions.Empty() {
+		var ext uint16
+		var extData cryptobyte.String
+		if !extensions.ReadUint16(&ext) ||
+			!extensions.ReadUint16LengthPrefixed(&extData) {
+			return false
+		}
+		if ok := proc(ext, extData); !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// splitClientHelloExtensions interprets data as a ClientHello message and
+// returns two strings: the first contains the start of the ClientHello up to
+// the start of the extensions; and the second is the length-prefixed
+// extensions. It also returns a bit indicating whether parsing succeeded.
+func splitClientHelloExtensions(data []byte) ([]byte, []byte, bool) {
+	s := cryptobyte.String(data)
+
+	var ignored uint16
+	var t cryptobyte.String
+	if !s.Skip(4) || // message type and uint24 length field
+		!s.ReadUint16(&ignored) || !s.Skip(32) || // vers, random
+		!s.ReadUint8LengthPrefixed(&t) { // session_id
+		return nil, nil, false
+	}
+
+	if !s.ReadUint16LengthPrefixed(&t) { // cipher_suites
+		return nil, nil, false
+	}
+
+	if !s.ReadUint8LengthPrefixed(&t) { // compression_methods
+		return nil, nil, false
+	}
+
+	return data[:len(data)-len(s)], s, true
+}
+
+func (c *Config) echSelectConfig() *ECHConfig {
+	for _, echConfig := range c.ClientECHConfigs {
+		// A suitable configuration is one that offers an HPKE ciphersuite that
+		// is supported by the client and indicates the version of ECH
+		// implemented by this TLS client.
+		if echConfig.isSupported() && extensionECH == echConfig.version {
+			return &echConfig
+		}
+	}
+	return nil
+}
+
+func (c *Config) echCanOffer() bool {
+	if c == nil {
+		return false
+	}
+	return c.ECHEnabled && c.echSelectConfig() != nil && c.MaxVersion >= VersionTLS13
```

Comment:
```suggestion
	return c.ECHEnabled && c.echSelectConfig() != nil && (c.MaxVersion >= VersionTLS13 || c.MaxVersion == 0)
```

Reply:
Changed to `c.maxSupportedVersion >= VersionTLS13`.